File: 73/set-matrix-zeroes.py
# 陣列最大是200x200，如果裡面有0，就把對應的row,col都設成0
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        zeroI, zeroJ = [], []
        M, N = len(matrix), len(matrix[0])
        for i in range(M):
            for j in range(N):
                if matrix[i][j]==0: # 遇到0時
                    zeroI.append(i) # 記錄對應的i
                    zeroJ.append(j) # 記錄對應的j

        for k in range(len(zeroI)): # 把記錄下來的i,j都逐一取出來
            i, j = zeroI[k], zeroJ[k]
            for ii in range(M): # 把整行 matrix[*][j] 全設為0
                matrix[ii][j]=0
            for jj in range(N): # 把整行 matrix[i][*] 全設為0
                matrix[i][jj]=0

File: 73/set-matrix-zeroes.cpp
// LeetCode 73. Set Matrix Zeroes
// 直覺寫出來的答案「是錯的」,因為0蔓延之後,變太多0了
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int M = matrix.size(), N = matrix[0].size();
        vector<int> up(N); //宣告一個陣列, 是放在上面up、用來「打勾勾」標注「有哪幾直條要刪」
        vector<int> left(M); //宣告一個陣列, 是放在左邊left、用來「打勾勾」標注「有哪幾橫條要刪」
        for(int i=0; i<M; i++){
            for(int j=0; j<N; j++){
                if(matrix[i][j]==0){ //遇到0的話, 要標注 left[i] 和 up[j]
                    up[j] = 1;
                    left[i] = 1;
                }
            }
        }
        for(int i=0; i<M; i++){ //最後再把它變成0
            if(left[i]==1) for(int jj=0; jj<N; jj++) matrix[i][jj]=0; //如果左邊有「打勾勾」,就整條刪 橫條
        }
        for(int j=0; j<N; j++){
            if(up[j]==1) for(int ii=0; ii<M; ii++) matrix[ii][j]=0; //如果上面有「打勾勾」,就整條刪 直條
        }
    }
};
